### PR TITLE
Pgsql

### PR DESCRIPTION
**Pull request by ksumag**
add pqsql exporter  container launch with scrape metrics config for Prometheus

### CODE REVIEW

File: tasks/postgres.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+      {{ monitoring_pgsql_parameters }}
+    ports:
+      - "{{ monitoring_pgsql_db_port }}:{{ monitoring_pgsql_db_port }}"
+    networks:
+      - name: "{{ monitoring_docker_network_name }}"
+        aliases:
+          - pg_exporter
+    restart_policy: always
+  when: monitoring_postgresql_enabled | bool
+  tags: ['pg_exporter']
```
Comment (dnillovna) [inline tasks/postgres.yml:15]:
```suggestion
  tags: ['pg_exporter']

```

File: tasks/postgres.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+- name: Run postresql_exporter via docker
+  docker_container:
+    name: monitoring_postgresql
+    image: "quay.io/prometheuscommunity/postgres-exporter"
+    env:
+      {{ monitoring_pgsql_parameters }}
```
Comment (dnillovna) [inline tasks/postgres.yml:6]:
```suggestion
    env: {{ monitoring_pgsql_parameters | default ({}) }}
```

File: tasks/postgres.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    ports:
+      - "{{ monitoring_pgsql_db_port }}:{{ monitoring_pgsql_db_port }}"
+    networks:
+      - name: "{{ monitoring_docker_network_name }}"
```
Comment (dnillovna) [inline tasks/postgres.yml:10]: What is the default value for the variable?

File: templates/prometheus.yml.j2 [hunk 4]
```diff
@@ -97,3 +97,11 @@ scrape_configs:
...
         replacement: blackbox:{{ monitoring_blackbox_port }}
 {% endif %}
+
+{% if monitoring_postgresql_enabled | default(true) %}
+  - job_name: "postgresql"
+    static_configs:
+      - targets: ['{{monitoring_pgsql_db_host}}:{{monitoring_pgsql_exporter_port}}']
+        labels:
+          hostname: {{monitoring_pgsql_db_host}}
+{% endif %}
```
Comment (dnillovna) [inline templates/prometheus.yml.j2:107]:
```suggestion
{% endif %}

```

File: templates/prometheus.yml.j2 [hunk 5]
```diff
@@ -97,3 +97,11 @@ scrape_configs:
       - target_label: __address__
         replacement: blackbox:{{ monitoring_blackbox_port }}
 {% endif %}
+
+{% if monitoring_postgresql_enabled | default(true) %}
+  - job_name: "postgresql"
+    static_configs:
+      - targets: ['{{monitoring_pgsql_db_host}}:{{monitoring_pgsql_exporter_port}}']
```
Comment (dnillovna) [inline templates/prometheus.yml.j2:104]:
```suggestion
      - targets: ['{{ monitoring_pgsql_db_host }}:{{ monitoring_pgsql_exporter_port }}']
```
do not forget to run linter)

File: templates/prometheus.yml.j2 [hunk 6]
```diff
@@ -97,3 +97,11 @@ scrape_configs:
...
+        labels:
+          hostname: {{monitoring_pgsql_db_host}}
```
Comment (dnillovna) [inline templates/prometheus.yml.j2:106]:
```suggestion
          hostname: {{ monitoring_pgsql_db_host }}
```

**Review by dnillovna — CHANGES_REQUESTED** [review]

File: tasks/postgres.yml (hunk 3 above)
Comment (ksumag) [reply tasks/postgres.yml:10]: monitoring, https://github.com/datagrok-ai/ansible-role-datagrok-monitoring/blob/pgsql/defaults/main.yml#L7

File: tasks/postgres.yml (hunk 2 above)
Comment (ksumag) [reply tasks/postgres.yml]: env: "{{ monitoring_pgsql_parameters | default ({}) }}"

File: defaults/main.yml [hunk 7]
```diff
@@ -39,6 +39,7 @@ monitoring_grafana_enabled: true
 monitoring_alertmanager_enabled: true
 monitoring_blackbox_enabled: true
 monitoring_cadvisor_enabled: true
+monitoring_postgresql_enabled: true
```
Comment (dnillovna) [inline defaults/main.yml:42]:
```suggestion
monitoring_pgsql_exporter_enabled: true
```
The names of the variables should be consistent. Check all the variables and make sure that they have similar prefixes, for example `monitoring_pgsql_exporter_`

File: tasks/main.yml [hunk 8]
```diff
@@ -5,6 +5,12 @@
   when:
     - monitoring_targets_inventory_group in group_names
 
+- name: include pgsql_monitoring
+  ansible.builtin.include_tasks:
+    file: postgres.yml
+  when:
+    - monitoring_targets_inventory_group in group_names
```
Comment (dnillovna) [inline tasks/main.yml:12]:
```suggestion
  delegate_to: "{{ monitoring_pgsql_db_host }}"
  run_once: true
```

File: tasks/postgres.yml [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    image: "quay.io/prometheuscommunity/postgres-exporter"
+    env: "{{ monitoring_pgsql_parameters | default ({}) }}"
+    ports:
+      - "{{ monitoring_pgsql_exporter_port }}:{{ monitoring_pgsql_exporter_port }}"
+    networks:
+      - name: "{{ monitoring_docker_network_name }}"
+        aliases:
+          - pg_exporter
+    restart_policy: always
+  delegate_to: "{{ monitoring_pgsql_db_host }}"
```
Comment (dnillovna) [inline tasks/postgres.yml:13]:
```suggestion
```

**Review by dnillovna — CHANGES_REQUESTED** [review]

File: templates/prometheus.yml.j2 [hunk 10]
```diff
@@ -97,3 +97,11 @@ scrape_configs:
       - target_label: __address__
         replacement: blackbox:{{ monitoring_blackbox_port }}
 {% endif %}
+
+{% if monitoring_postgresql_enabled | default(true) %}
+  - job_name: "postgresql"
+    static_configs:
+      - targets: ['{{ hostvars[monitoring_pgsql_db_host]["ansible_host"] }}:{{ monitoring_pgsql_exporter_port }}']
```
[the remark this answers is not in the document]
Comment (ksumag) [reply templates/prometheus.yml.j2:104]: don't understand link on the same PR, I've got one target to scrape, what's wrong?

[the remark this answers is not in the document]
Comment (dnillovna) [reply templates/prometheus.yml.j2:104]:
```
      - targets: [
        {%- if host == inventory_hostname -%}
          'cadvisor:{{ monitoring_cadvisor_port }}'
        {%- else -%}
          '{{ hostvars[host]["monitoring_advertise_address"] | default(hostvars[host]["ansible_host"]) }}:{{ monitoring_cadvisor_port }}'
        {%- endif -%}]
```

File: templates/prometheus.yml.j2 [hunk 11]
```diff
@@ -97,3 +97,11 @@ scrape_configs:
...
         replacement: blackbox:{{ monitoring_blackbox_port }}
 {% endif %}
+
+{% if monitoring_pgsql_exporter_enabled | default(true) %}
+  - job_name: "postgresql"
+    static_configs:
+      - targets: ['{{ hostvars[monitoring_pgsql_db_host]["ansible_host"] }}:{{ monitoring_pgsql_exporter_port }}']
+        labels:
+          hostname: {{ monitoring_pgsql_db_host }}
+{% endif %}
```
Comment (dnillovna) [inline templates/prometheus.yml.j2:107]:
```
    relabel_configs:
      - source_labels: [hostname]
        target_label: instance
```

**Review by dnillovna — CHANGES_REQUESTED** [review]

File: templates/alertmanager.yml.j2 [hunk 12]
```diff
@@ -20,11 +21,17 @@ global:
...
   receiver: 'DevOps'
-  group_by: [alertname, InstanceState]
+  group_by: [alertname, InstanceState, name]
   group_interval: {{ monitoring_alerts_group_message_interval }}
   repeat_interval: {{ monitoring_alerts_repeat_message_interval }}
+  routes: 
+  - receiver: 'null'
+    matchers:
+    - name!~"^(monitoring|dev|test|demo|registry|release|localhost_).*$"
+    - job="container_metrics"
```
Comment (dnillovna) [inline templates/alertmanager.yml.j2:31]: I should not be enabled by default. Cause some of our clients won't have such container names.
I suggest
```suggestion
{% if monitoring_except_rules | default([]) | length > 0 %}
  routes: 
  {% for rule in monitoring_except_rules %}
  - receiver: "{{ rule.receiver | default('null') }}"
    matchers:
    {% for matcher in  rule.matchers %}
        - {{ matcher }}
    {% endfor %}
{% endfor %}
```

File: templates/alert_rules.yml.j2 [hunk 13]
```diff
@@ -44,110 +44,113 @@ groups:
...
-      expr: (sum(rate(container_cpu_usage_seconds_total{name!=""}[3m])) BY (instance, name) * 100) > 90
-      for: 3m
+    - alert: "ContainerCpuUsage is above 90%"
+      expr: (sum(rate(container_cpu_usage_seconds_total{name!=""}[3m])) BY (instance, name, job) * 100) > 90
+      for: 10m
       labels:
         severity: warning
       annotations:
-        summary: Container CPU usage (instance {{ $labels.instance }})
+        summary: "{{ $labels.instance }}: Container {{ $labels.name }} CPU usage > 90%"
```
Comment (dnillovna) [inline templates/alert_rules.yml.j2:109]:
```suggestion
        summary: "{{ $labels.instance }}: {{ $labels.name }} container CPU usage > 90%"
```

File: templates/alert_rules.yml.j2 [hunk 14]
```diff
@@ -44,110 +44,113 @@ groups:
...
-      for: 2m
+    - alert: "ContainerMemoryUsage is above 85%"
+      expr: (sum(container_memory_working_set_bytes{name!=""}) BY (instance, name, job) / sum(container_spec_memory_limit_bytes > 0) BY (instance, name, job) * 100) > 85
+      for: 5m
       labels:
         severity: warning
       annotations:
-        summary: Container Memory usage (instance {{ $labels.instance }})
-        description: "Container Memory usage is above 86%\n  VALUE = {{ $value }}\n  LABELS = {{ $labels }}"
+        summary: "{{ $labels.instance }}: Container {{ $labels.name }} Memory usage > 85%"
```
Comment (dnillovna) [inline templates/alert_rules.yml.j2:118]:
```suggestion
        summary: "{{ $labels.instance }}: {{ $labels.name }} container Memory usage > 85%"
```

File: templates/alert_rules.yml.j2 [hunk 15]
```diff
@@ -44,110 +44,113 @@ groups:
...
 
     - alert: ContainerHighThrottleRate
       expr: rate(container_cpu_cfs_throttled_seconds_total[3m]) > 1
-      for: 2m
+      for: 5m
       labels:
         severity: warning
       annotations:
-        summary: Container high throttle rate (instance {{ $labels.instance }})
+        summary: "{{ $labels.instance }}: Container {{ $labels.name }} high throttle rate"
```
Comment (dnillovna) [inline templates/alert_rules.yml.j2:127]:
```suggestion
        summary: "{{ $labels.instance }}: {{ $labels.name }} container high throttle rate"
```

File: templates/alert_rules.yml.j2 [hunk 16]
```diff
@@ -44,110 +44,113 @@ groups:
...
+
     - alert: BlackboxProbeFailed
       expr: probe_success == 0
-      for: 1m
+      for: 2m
       labels:
         severity: critical
       annotations:
-        summary: Blackbox probe failed (instance {{ $labels.instance }})
+        summary: "Blackbox probe failed on {{ $labels.instance }}"
```
Comment (dnillovna) [inline templates/alert_rules.yml.j2:136]:
```suggestion
        summary: "{{ $labels.instance }}: Blackbox probe failed"
```

**Review by dnillovna — CHANGES_REQUESTED** [review]

File: templates/alert_rules.yml.j2 [hunk 17]
```diff
@@ -44,110 +44,113 @@ groups:
...
+
     - alert: BlackboxProbeHttpFailure
       expr: probe_http_status_code <= 199 OR probe_http_status_code >= 400
-      for: 0m
+      for: 5m
       labels:
         severity: critical
       annotations:
-        summary: Blackbox probe HTTP failure (instance {{ $labels.instance }})
+        summary: "{{ $value }} HTTP code on {{ $labels.instance }}"
```
Comment (dnillovna) [inline templates/alert_rules.yml.j2:145]:
```suggestion
        summary: "{{ $labels.instance }}: {{ $value }} HTTP code"
```

**Pull request merged**